### PR TITLE
BUG: remove usage of _constructor._get_axis_number (fix when _constructor properties return callables)

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
See report at https://github.com/pandas-dev/pandas/issues/32860#issuecomment-697993089 (and we also run into this one case in geopandas). Also related to https://github.com/pandas-dev/pandas/issues/32638

As far as I know this is the only remaining case where we rely on `_constructor`/`_constructor_sliced`/`_constructor_expanddim` returning a class and not a generic callable (in geopandas, we also have a workaround for this one case). 

This specific case was introduced in https://github.com/pandas-dev/pandas/pull/34837 to avoid circular dependencies/imports, so therefore I added the DataFrame import inline.

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
Found one more, in `replace()` which was introduced in https://github.com/pandas-dev/pandas/pull/39152 (restored the inline import that was there before that PR)

**Review by jreback — CHANGES_REQUESTED** [review]
no objection, can you add a whatsnew

**Comment by jreback** [comment]
tests are failing as well :->

**Comment by jorisvandenbossche** [comment]
Ah, that's another case, but where we didn't call a method on it, but used in an `issubclass` check.

File: pandas/core/reshape/concat.py [hunk 1]
```diff
@@ -539,7 +542,7 @@ def __init__(
         self.new_axes = self._get_new_axes()
 
     def get_result(self):
-        cons: type[DataFrame | Series]
+        cons: Callable
```
Comment (jbrockmendel) [inline pandas/core/reshape/concat.py:545]: could annotate as Callable returning DataFrame|Series

Comment (jorisvandenbossche) [reply pandas/core/reshape/concat.py:545]: The `Callable` typing is actually giving errors .. (with or without return type)

Comment (jorisvandenbossche) [reply pandas/core/reshape/concat.py:545]: Ah, but I probably have to update the actual properties' annotation to callable as well

**Comment by jorisvandenbossche** [comment]
Are people OK with considering this for 1.4.x? (the actual change is quite small, the bulk of the diff is due to typing)

This is actually fixing a regression for geopandas, although not a regression in 1.4.0, but somewhere between 1.0 and 1.4.

File: pandas/_testing/__init__.py [hunk 2]
```diff
@@ -803,23 +803,23 @@ class SubclassedSeries(Series):
 
     @property
     def _constructor(self):
-        return SubclassedSeries
+        return lambda *args, **kwargs: SubclassedSeries(*args, **kwargs)
```
Comment (jbrockmendel) [inline pandas/_testing/__init__.py:806]: can you add a comment to one of these pointing back to the relevant issue

File: pandas/core/generic.py [hunk 3]
```diff
@@ -5922,11 +5910,7 @@ def astype(
         # GH 19920: retain column metadata after concat
         result = concat(results, axis=1, copy=False)
         # GH#40810 retain subclass
-        # Incompatible types in assignment (expression has type "NDFrameT",
-        # variable has type "DataFrame")
-        # Argument 1 to "NDFrame" has incompatible type "DataFrame"; expected
-        # "Union[ArrayManager, SingleArrayManager, BlockManager, SingleBlockManager]"
-        result = self._constructor(result)  # type: ignore[arg-type,assignment]
+        result = self._constructor(result)  # type: ignore[assignment]
```
Comment (jbrockmendel) [inline pandas/core/generic.py:5913]: can you add the mypy complaint as a comment in the line above

File: pandas/_testing/__init__.py (hunk 2 above)
Comment (jorisvandenbossche) [reply pandas/_testing/__init__.py:806]: Yes, done

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
@meeseeksdev backport 1.4.x

**Comment by simonjayhawkins** [comment]
The auto backport failed as #45272 is not backported. The changes here to that code were only the mypy comments. Those changes have therefore been omitted in the manual backport.

**Pull request merged**